Add Hash Table implementation using chaining with Linked Lists

hellovietduc opened the pull request:
Closes #1

Commit:
Add Hash Table implementation

trungducc commented on hash_table/hash_table.go:
```diff
@@ -0,0 +1,134 @@
+
+func NewHashTable() *HashTable {
+	return &HashTable{
+		arr: make([]*Node, 10),
```
I prefer to use `8` or `16` as default capacity instead of `10`

hellovietduc replied:
Is it because it's a power of 2?

trungducc replied:
Yes, we should use a power of 2 so that we can control what capacity will be

trungducc commented on hash_table/hash_table.go:
```diff
@@ -0,0 +1,134 @@
+
+func (h *HashTable) Insert(key int, value int) {
+	// same load factor as Java 10
+	if h.getLoadFactor() >= 0.75 {
```
Please define constant variable instead of using directly.

trungducc commented on hash_table/hash_table.go:
```diff
@@ -0,0 +1,134 @@
+	node.next = &Node{
+		key:   key,
+		value: value,
+	}
```
I prefer to creating a new node before using to reduce duplicate code.

trungducc commented on hash_table/hash_table.go:
```diff
@@ -0,0 +1,134 @@
+	return float64(h.Size()) / float64(len(h.arr))
+}
+
+func (h *HashTable) getHash(key int) int {
```
Function name should be `hash` instead of `getHash`

trungducc commented on hash_table/hash_table.go:
```diff
@@ -0,0 +1,134 @@
+
+}
+
+func (h *HashTable) Search(key int) (int, error) {
```
Search function will return a value if the key exists. Otherwise, it can return `nil`.

hellovietduc replied:
Unfortunately, we can't return `nil` in place of an integer in Go. The convention is to return another value, which is `error` in this case to determine if the operation succeeds or not.

trungducc replied:
I'm not sure about Go part but normally `search` function should not give an error. @thanks173  How do you think about this issue?

thanks173 replied:
Default implementation of Go (using `map`) will just return `nil` or zero value (depending on the value type).
I agree with @trungducc

hellovietduc replied:
`map` in Go is dynamically typed, keys and values can be of any types. For the sake of simplicity, I'm implementing a Hash Table of Integers, so there's no such thing as `nil`. If I return `-1` then I wouldn't know if it's `nil` or a real `-1` value.

trungducc replied:
It's ok to return a number if the key doesn't exist. There are 2 points here:
- This function should not give an error
- If the key doesn't exist, I prefer to return min value of Int and define it as a public constant.  For example, we can use `search(1) == HashTable.NOT_FOUND` to determine the key exist or not.

hellovietduc replied:
I think I need help from @thanks173 here. I'm still learning Go so I don't know if there's a convention for declaring enums, or constants in Go.

thanks173 replied:
In the default implementation, Go use a two-value assignment to check if a key is present or not, and it's optional.
```
elem, ok = m[not_existed_key] // return nil/0/false for elem, and false for ok
elem = m[not_existed_key] // return nil/0/false for elem
```
For the sake of simplicity, I think returning `0` value is ok. It's the same with initializing an integer without giving it any value.

thanks173 replied:
For enums, you could use Go's equivelent `iota`. For the convention, you can read [here](https://yourbasic.org/golang/iota/#iota-basic-example). 

But as the function should return `int`, I don't think it's a good idea to use another `int` to determine the key exist or not.

The zero value (`nil`, `0`, `false`, `""`,...), on the other hand, is actually more useful than you may think. You can read it [here](https://dave.cheney.net/2013/01/19/what-is-the-zero-value-and-why-is-it-useful). In the end (your application), you are dealing with the value stored in the hash table, not the table itself. So using the zero value as a known default is always recommended (i.e, checking hash table for a logged-in user, the value `0` should mean that the user is not logged-in, whether the hash table contains that `user_id` or not).

Your approach is fine btw, it's just complicated.

trungducc commented on hash_table/hash_table.go:
```diff
@@ -0,0 +1,134 @@
+	return node.value, nil
+}
+
+func (h *HashTable) Delete(key int) (int, error) {
```
Delete function also doesn't throw error if key can't be found.

hellovietduc replied:
My intention is to allow users of this struct to know whether the operation succeeds or not.

trungducc replied:
After calling this function, there is no value with given key in hash table. I think this function will always success even the key exists or not.

thanks173 commented on hash_table/hash_table.go:
```diff
@@ -0,0 +1,134 @@
+	}
+}
+
+func (h *HashTable) Insert(key int, value int) {
```
Missing "insert with existed key" case (we should overwrite the value)

trungducc replied:
@thanks173 nice catch. I suggest to request changes instead of commenting for the next time you find something needs to improve.

Commit:
change struct name to hashmap

hellovietduc commented:
I renamed the struct to `HashMap`, because I'm planning to implement `HashSet` as well. Also corrected all the errors pointed out from you guys' reviews @trungducc  @thanks173.

Commit:
add comments for exported elements

trungducc commented on hashtable/hashmap.go:
```diff
@@ -0,0 +1,161 @@
+const (
+	// KeyNotFound is a constant indicates that
+	// a key cannot be found in the HashMap.
+	KeyNotFound     = 0
```
`KeyNotFound` should be a rarely used value. I prefer to using min value of `int` type

trungducc commented on hashtable/hashmap.go:
```diff
@@ -0,0 +1,161 @@
+		curNode = curNode.next
+	}
+
+	return curNode.value
```
Looks like the code is duplicated. There are some pieces of code which follow same steps:
- Check if node is not nil
- Iterate nodes
- Do something with the last node

Could you do it without making duplicated code?

Commit:
refactor for loops in linked lists

hellovietduc commented:
Anymore reviews? @trungducc @thanks173 ?

thanks173 reviewed (changes requested):
Only increase size when inserting new key

trungducc commented:
@vietduc01100001 Did you remove logging from test cases? BTW, you can make `insert`, `search`, `delete` shorter by creating a private search method which returns an index.  When you are inserting, index will determine empty, deleted slot or slot which same key. `search` function will return `true` if key is valid.

thanks173 commented on hashtable/hashmap.go:
```diff
@@ -0,0 +1,147 @@
+	}
+
+	// remove the reference from this node
+	curNode.next = nil
```
I don't understand what you're trying to do here.
Btw these 2 lines can never be executed. You forgot to break the loop.

hellovietduc replied:
I fixed that in a new commit

Commit:
move to chaining subfolder

Commit:
refactor rehash function

Commit:
add few comments

Commit:
fix delete not break loop

Commit:
check capacity in tests

This reverts commit 75d61020ea343555bf0bde08f15d3fd33a5f3055.

thanks173 commented on hashtable/chaining/hashmap.go:
```diff
@@ -0,0 +1,154 @@
+		curNode = curNode.next
+	}
+
+	if curNode == nil {
```
After the loop, if the key is not existed, the `curNode` will always be `nil`

thanks173 commented on hashtable/chaining/hashmap.go:
```diff
@@ -0,0 +1,154 @@
+		value: value,
+	}
+
+	h.size++
```
Only increasing size when adding new key to the table

thanks173 commented:
Just my personal preference, but I would avoid using `git revert`, because it will leave something I don't want anyone to see in `git log`. Even worse, if the reverted commit contains some sensitive informations (sensitive informations shouldn't be on git at all), `git revert` will not delete them at all, and those sensitive informations will probably be stored forever in the log.

This can be fixed by using "squash and merge", but I personally would delete the remote branch and checkout the local branch to the earlier commit.

Again, this's just my personal preference and something I want to share.

hellovietduc commented:
Can you explain on the "delete remote branch and checkout local branch to earlier commit" part? I did squash rebase before but this time I just want to try revert, because I've never done it.

thanks173 commented:
> Can you explain on the "delete remote branch and checkout local branch to earlier commit" part? I did squash rebase before but this time I just want to try revert, because I've never done it.

You delete remote branch in github. In your computer, you checkout the local branch to the earlier commit, then push that commit back to a new remote branch in github.

hellovietduc commented:
Then that new branch still links to the PR?

Commit:
refactor to use doubly linked lists

Commit:
fix reviewed cases

trungducc commented on hashtable/chaining/hashmap.go:
```diff
@@ -0,0 +1,154 @@
+func (h *HashMap) Delete(key int) {
+	if h.getLoadFactor() < lowerLoadFactor {
+		h.shrink()
+	}
```
Should we only shrink if deleting successfully?

trungducc replied:
@vietduc01100001 Did you forget to check this part?

hellovietduc replied:
I did move that down after the `foundNode == nil` check.

trungducc commented on hashtable/chaining/hashmap.go:
```diff
@@ -0,0 +1,154 @@
+
+	if foundNode.next == nil {
+		// last node in the linked list
+		foundNode.prev.next = nil
```
Did you forget calling `return` here?

trungducc commented on hashtable/chaining/hashmap.go:
```diff
@@ -0,0 +1,154 @@
+func (h *HashMap) rehash(newLength int) {
+	newArr := make([]*node, newLength)
+	oldArr := h.items
+	h.items = newArr
```
I think you can remove line 143 and change line 145 to `h.items = make([]*node, newLength)`

trungducc commented on hashtable/chaining/hashmap_test.go:
```diff
@@ -0,0 +1,69 @@
+	"testing"
+)
+
+func TestHashMap(t *testing.T) {
```
Not sure about how test cases are written in Go but in other languages, each method will test only 1 test case.

hellovietduc replied:
I'm not sure neither. I'm assuming that the `TestHashMap` is like a `describe` function, and `t.Run` is like small test cases. How about this @thanks173 ?

trungducc replied:
@thanks173 Do you have any suggestion?

trungducc commented on hashtable/chaining/hashmap.go:
```diff
@@ -0,0 +1,154 @@
+		}
+		curNode.next = newNode
+		newNode.prev = curNode
+	}
```
Just a suggestion but if you use sentinel node and `searchNode` returns a suitable node, you won't have to search again in this case.

trungducc reviewed (changes requested):
Looks great. I am going to merge this PR after conversations are resolved.

trungducc commented on .github/workflows/go.yml:
```diff
@@ -0,0 +1,30 @@
+    - name: Check out code into the Go module directory
+      uses: actions/checkout@v2
+
+    - name: Test
```
@vietduc01100001 Will you remove this file?
